contract evaluateTransaction doesn't throw an error if the chaincode responds with an error

Using node.js chaincode and node.js thin client I execute a submit transaction where the transaction name doesn't exist and the contract api returns an error informing you of the fact. the submit throws an error and the error provides appropriate detail
```
Error: 10 ABORTED: failed to endorse transaction: [{address:"peer0.org1.example.com:7051" msp_id:"Org1MSP" message:"error 500, error in simulation: transaction returned with failure: Error: You've asked to invoke a function that does not exist: put" }]
```

Doing the same using evaluate transaction on a contract results in an empty response, no error is thrown and the contract definitely reports that the function name doesn't exist


```
2021-09-14 16:04:06.481 UTC 00ec ERRO [endorser] simulateProposal -> failed to invoke chaincode basic, error: transaction returned with failure: Error: You've asked to invoke a function that does not exist: ReadAsset1
github.com/hyperledger/fabric/core/chaincode.processChaincodeExecutionResult
        /go/src/github.com/hyperledger/fabric/core/chaincode/chaincode_support.go:188
github.com/hyperledger/fabric/core/chaincode.(*ChaincodeSupport).Execute
        /go/src/github.com/hyperledger/fabric/core/chaincode/chaincode_support.go:162
github.com/hyperledger/fabric/core/endorser.(*SupportImpl).Execute
        /go/src/github.com/hyperledger/fabric/core/endorser/support.go:126
github.com/hyperledger/fabric/core/endorser.(*Endorser).callChaincode
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:119
github.com/hyperledger/fabric/core/endorser.(*Endorser).simulateProposal
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:186
github.com/hyperledger/fabric/core/endorser.(*Endorser).ProcessProposalSuccessfullyOrError
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:402
github.com/hyperledger/fabric/core/endorser.(*Endorser).ProcessProposal
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:344
github.com/hyperledger/fabric/internal/pkg/gateway.(*EndorserServerAdapter).ProcessProposal
        /go/src/github.com/hyperledger/fabric/internal/pkg/gateway/gateway.go:38
github.com/hyperledger/fabric/internal/pkg/gateway.(*Server).Evaluate
        /go/src/github.com/hyperledger/fabric/internal/pkg/gateway/api.go:66
github.com/hyperledger/fabric-protos-go/gateway._Gateway_Evaluate_Handler.func1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/hyperledger/fabric-protos-go/gateway/gateway.pb.go:1185
github.com/hyperledger/fabric/internal/peer/node.unaryGrpcLimiter.func1
        /go/src/github.com/hyperledger/fabric/internal/peer/node/grpc_limiters.go:44
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/hyperledger/fabric/common/grpclogging.UnaryServerInterceptor.func1
        /go/src/github.com/hyperledger/fabric/common/grpclogging/server.go:92
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/hyperledger/fabric/common/grpcmetrics.UnaryServerInterceptor.func1
        /go/src/github.com/hyperledger/fabric/common/grpcmetrics/interceptor.go:31
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34
github.com/hyperledger/fabric-protos-go/gateway._Gateway_Evaluate_Handler
        /go/src/github.com/hyperledger/fabric/vendor/github.com/hyperledger/fabric-protos-go/gateway/gateway.pb.go:1187
google.golang.org/grpc.(*Server).processUnaryRPC
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:1180
google.golang.org/grpc.(*Server).handleStream
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:1503
google.golang.org/grpc.(*Server).serveStreams.func1.2
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:843
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1371 channel=mychannel txID=611d2cc7
2021-09-14 16:04:06.481 UTC 00ed WARN [endorser] ProcessProposal -> Failed to invoke chaincode channel=mychannel chaincode=basic error="error in simulation: transaction returned with failure: Error: You've asked to invoke a function that does not exist: ReadAsset1"
2021-09-14 16:04:06.481 UTC 00ee INFO [comm.grpc.server] 1 -> unary call completed grpc.service=gateway.Gateway grpc.method=Evaluate grpc.peer_address=172.21.0.1:34772 grpc.code=OK grpc.call_duration=5.8841ms
```

Just to confirm, the same thing happens with the Go Client and is actually worse as the application crashes
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x8b31a1]

goroutine 1 [running]:
github.com/hyperledger/fabric-gateway/pkg/client.(*Proposal).Evaluate(0xc000572c00, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/dave/github-mine/fabric-gateway/pkg/client/proposal.go:100 +0x201
github.com/hyperledger/fabric-gateway/pkg/client.(*Contract).Evaluate(0xc00052be28, 0x9d8610, 0x3, 0xc00052bd40, 0x1, 0x1, 0x0, 0x0, 0x2c, 0xc00063a270, ...)
        /home/dave/github-mine/fabric-gateway/pkg/client/contract.go:71 +0xb6
github.com/hyperledger/fabric-gateway/pkg/client.(*Contract).EvaluateTransaction(0xc00052be28, 0x9d8610, 0x3, 0xc00052bdf0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
        /home/dave/github-mine/fabric-gateway/pkg/client/contract.go:59 +0xa5
main.exampleSubmit(0xc00000ec18)
        /home/dave/github-mine/fabric-gateway/samples/go/sample.go:87 +0x290
main.main()
        /home/dave/github-mine/fabric-gateway/samples/go/sample.go:47 +0x1b9
```

--- a/internal/pkg/gateway/api.go
+++ b/internal/pkg/gateway/api.go
@@ -66,20 +66,22 @@ func (gs *Server) Evaluate(ctx context.Context, request *gp.EvaluateRequest) (*g
 
 	response, err := endorser.client.ProcessProposal(ctx, signedProposal)
 	if err != nil {
-		logger.Debugw("Evaluate call to endorser failed", "channel", request.ChannelId, "txid", request.TransactionId, "endorserAddress", endorser.endpointConfig.address, "endorserMspid", endorser.endpointConfig.mspid, "error", err)
+		logger.Debugw("Evaluate call to endorser failed", "chaincode", chaincodeID, "channel", channel, "txid", request.TransactionId, "endorserAddress", endorser.endpointConfig.address, "endorserMspid", endorser.endpointConfig.mspid, "error", err)
 		return nil, wrappedRpcError(err, "failed to evaluate transaction", endpointError(endorser.endpointConfig, err))
 	}
 
-	retVal, err := getTransactionResponse(response)
-	if err != nil {
-		logger.Debugw("Evaluate call to endorser returned a malformed or error response", "channel", request.ChannelId, "txid", request.TransactionId, "endorserAddress", endorser.endpointConfig.address, "endorserMspid", endorser.endpointConfig.mspid, "error", err)
-		return nil, rpcError(codes.Unknown, err.Error(), endpointError(endorser.endpointConfig, err))
-	}
-	evaluateResponse := &gp.EvaluateResponse{
-		Result: retVal,
+	rr := response.GetResponse()
+	if rr != nil && (rr.Status < 200 || rr.Status >= 400) {
+		logger.Debugw("Evaluate call to endorser returned a malformed or error response", "chaincode", chaincodeID, "channel", channel, "txid", request.TransactionId, "endorserAddress", endorser.endpointConfig.address, "endorserMspid", endorser.endpointConfig.mspid, "status", rr.Status, "message", rr.Message)
+		err = fmt.Errorf("error %d returned from chaincode %s on channel %s: %s", rr.Status, chaincodeID, channel, rr.Message)
+		return nil, rpcError(codes.Unknown, "error returned from chaincode: "+rr.Message, endpointError(endorser.endpointConfig, err))
 	}
 
-	logger.Debugw("Evaluate call to endorser returned success", "channel", request.ChannelId, "txid", request.TransactionId, "endorserAddress", endorser.endpointConfig.address, "endorserMspid", endorser.endpointConfig.mspid, "status", retVal.GetStatus(), "message", retVal.GetMessage())
+	evaluateResponse := &gp.EvaluateResponse{
+		Result: rr,
+	}
+
+	logger.Debugw("Evaluate call to endorser returned success", "channel", request.ChannelId, "txid", request.TransactionId, "endorserAddress", endorser.endpointConfig.address, "endorserMspid", endorser.endpointConfig.mspid, "status", rr.GetStatus(), "message", rr.GetMessage())
 	return evaluateResponse, nil
 }
 
@@ -247,13 +249,8 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
 		return nil, status.Errorf(codes.Aborted, "failed to assemble transaction: %s", err)
 	}
 
-	retVal, err := getTransactionResponse(responses[0])
-	if err != nil {
-		return nil, status.Errorf(codes.Aborted, "failed to extract transaction response: %s", err)
-	}
-
 	endorseResponse := &gp.EndorseResponse{
-		Result:              retVal,
+		Result:              responses[0].GetResponse(),
 		PreparedTransaction: env,
 	}
 	return endorseResponse, nil

--- a/internal/pkg/gateway/api_test.go
+++ b/internal/pkg/gateway/api_test.go
@@ -309,11 +309,11 @@ func TestEvaluate(t *testing.T) {
 				proposalResponseStatus:  400,
 				proposalResponseMessage: "Mock chaincode error",
 			},
-			errString: "rpc error: code = Unknown desc = error 400, Mock chaincode error",
+			errString: "rpc error: code = Unknown desc = error returned from chaincode: Mock chaincode error",
 			errDetails: []*pb.EndpointError{{
 				Address: "peer1:8051",
 				MspId:   "msp1",
-				Message: "error 400, Mock chaincode error",
+				Message: "error 400 returned from chaincode test_chaincode on channel test_channel: Mock chaincode error",
 			}},
 		},
 		{

--- a/internal/pkg/gateway/apiutils.go
+++ b/internal/pkg/gateway/apiutils.go
@@ -17,31 +17,6 @@ import (
 	"google.golang.org/grpc/status"
 )
 
-func getTransactionResponse(response *peer.ProposalResponse) (*peer.Response, error) {
-	var retVal *peer.Response
-
-	if response.GetPayload() != nil {
-		payload, err := protoutil.UnmarshalProposalResponsePayload(response.Payload)
-		if err != nil {
-			return nil, err
-		}
-
-		extension, err := protoutil.UnmarshalChaincodeAction(payload.Extension)
-		if err != nil {
-			return nil, err
-		}
-
-		if extension.GetResponse() != nil {
-			if extension.Response.Status < 200 || extension.Response.Status >= 400 {
-				return nil, fmt.Errorf("error %d, %s", extension.Response.Status, extension.Response.Message)
-			}
-			retVal = extension.Response
-		}
-	}
-
-	return retVal, nil
-}
-
 func getChannelAndChaincodeFromSignedProposal(signedProposal *peer.SignedProposal) (string, string, bool, error) {
 	if signedProposal == nil {
 		return "", "", false, fmt.Errorf("a signed proposal is required")

